### PR TITLE
Fix dist kvstore for trainer and flaky dist kvstore test

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
Previously there were no unit tests for trainer with dist kvstore. #11429 introduced a bug for dist kvstore in trainer, where it calls `pull(ignore_sparse=False)` which is NOT implemented for dist kvstore. This PR corrects the API call and adds unit test for it. Now `ignore_sparse` is set to False for non-distributed kvstore. 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by marcoabreu** [comment]
Hello Haibin,

please take note that integrationtest_ubuntu_gpu_dist_kvstore is currently disabled due to this test throwing exceptions (which strangely don't cause the run to fail). This is tracked at #11441

**Comment by ctcyang** [comment]
I see an error for `gluon_step`, but the other two pass. I ran `../../tools/launch.py -n 7 --launcher local python dist_sync_kvstore.py --type=gluon_step`. The error log is attached [dist_kvstore_error.txt](https://github.com/apache/incubator-mxnet/files/2182138/dist_kvstore_error.txt).

**Comment by ctcyang** [comment]
The three nightly tests now pass. LGTM

**Comment by eric-haibin-lin** [comment]
@marcoabreu I've fixed and enabled the test.

**Comment by marcoabreu** [comment]
Thank you. Did you investigate why there were errors thrown in the log but the test has not been marked as failed?

File: tests/nightly/dist_sync_kvstore.py [hunk 1]
```diff
@@ -388,8 +388,42 @@ def check_trainer_kv_type(stype, grad_stype, update_on_kv):
     check_trainer_kv_type('default', 'default', False)
     check_trainer_kv_type('default', 'row_sparse', True)
     check_trainer_kv_type('row_sparse', 'row_sparse', True)
-    print('worker ' + str(my_rank) + ' passed test_gluon_trainer')
-
+    print('worker ' + str(my_rank) + ' passed test_gluon_trainer_type')
+
+def test_gluon_trainer_step():
+    def check_trainer_step():
+        ctx = mx.cpu(0)
```
Comment (marcoabreu) [inline tests/nightly/dist_sync_kvstore.py]: is this on purpose? We run our KVstore tests on a GPU instance. if we don't require GPU, please downgrade to using a cpu instance.

File: ci/docker/runtime_functions.sh [hunk 2]
```diff
@@ -724,6 +724,17 @@ integrationtest_ubuntu_gpu_cpp_package() {
...
 
+integrationtest_ubuntu_cpu_dist_kvstore() {
+    set -ex
+    export PYTHONPATH=./python/
+    export MXNET_STORAGE_FALLBACK_LOG_VERBOSE=0
+    cd tests/nightly/
+    ../../tools/launch.py -n 7 --launcher local python dist_sync_kvstore.py --type=invalid
+    ../../tools/launch.py -n 7 --launcher local python dist_sync_kvstore.py --type=gluon_type
+    ../../tools/launch.py -n 7 --launcher local python dist_sync_kvstore.py --type=gluon_step
+    ../../tools/launch.py -n 7 --launcher local python dist_sync_kvstore.py --type=gluon_sparse_step
```
Comment (marcoabreu) [inline ci/docker/runtime_functions.sh]: Sorry, I'm having a hard time to determine which tests require a gpu and which ones dont. Would it be possible to add some way to distinguish the tests (like adding cpu and gpu to the names, separating them into different files or adding an argument).

File: Jenkinsfile [hunk 3]
```diff
@@ -979,20 +979,31 @@ try {
...
+          timeout(time: max_time, unit: 'MINUTES') {
+            init_git()
+            unpack_lib('gpu')
+            docker_run('ubuntu_gpu', 'integrationtest_ubuntu_gpu_dist_kvstore', true)
+            publish_test_coverage()
+          }
+        }
+      }
+    },
+    'dist-kvstore tests CPU': {
```
Comment (marcoabreu) [inline Jenkinsfile]: awesome!

**Comment by eric-haibin-lin** [comment]
Re: "Did you investigate why there were errors thrown in the log but the test has not been marked as failed?"

Yes, i found `init_kv()` was called twice and throwing that exception. Now it is called only once.

**Comment by marcoabreu** [comment]
Sure, but I mean why the script did not terminate with an error code. The exceptions were logged but the script did not actually fail with an error.

**Comment by eric-haibin-lin** [comment]
I thought @rahul003 fixed that? Could you confirm?

**Comment by eric-haibin-lin** [comment]
@rahul003 and I figured that the timeout was due to significant launch overhead of omp auto tuning. This is extremely slow when launching the test in local mode and all 15 processes are doing omp tuning, sharing the same instance. Creating kvstore takes 2 minutes when n=3.

**Comment by marcoabreu** [comment]
Awesome! Am I right in assuming that nobody would run this locally?

**Comment by eric-haibin-lin** [comment]
nobody except CI :)

**Comment by marcoabreu** [comment]
Okay excellent

File: tests/nightly/dist_sync_kvstore.py [hunk 4]
```diff
@@ -398,18 +432,25 @@ def check_trainer_kv_type(stype, grad_stype, update_on_kv):
...
     if opt.type == 'all' or opt.type == 'init':
         test_sync_init(opt.gpu)
+    kv = init_kv()
     if opt.type == 'all' or opt.type == 'default':
-        kv = init_kv()
         kv = set_optimizer(use_multiprecision=opt.multiprecision)
         test_sync_push_pull(opt.nrepeat)
-    # dont run non compressed tests after this as kvstore compression will be set here
+    # don't run non compressed tests after this as kvstore compression will be set here
     if opt.type == 'all' or opt.type == 'compressed':
```
Comment (rahul003) [inline tests/nightly/dist_sync_kvstore.py]: Could you remove this all parameter now?

Comment (eric-haibin-lin) [reply tests/nightly/dist_sync_kvstore.py]: why remove it?

Comment (rahul003) [reply tests/nightly/dist_sync_kvstore.py]: Because in the current script, it no longer represents all tests.

**Pull request merged**